### DISCUSSION TITLE
API

### DISCUSSION THREAD

**Discussion by atesz19** [Jikoo/OpenInv#38] [Q&A]
Hello!
I want to use this plugin to remove items from all offline players inventory and enderchest

Maybe can i get a little help with the api? I cant find any documentation :(

**Comment by atesz19** [comment]
if i trying to create a new OpenInv() then my plugin crashes with java.lang.IllegalArgumentException: Plugin already initialized!

**Comment by Jikoo** [comment]
You can obtain an instance of OpenInv through the plugin manager like any other plugin. You can cast that to `IOpenInv` and use accordingly.

**Comment by atesz19** [comment]
Thank you!
also i need a little help
ISpecialEnderChest ec = ((IOpenInv) plugin.getServer().getPluginManager().getPlugin("OpenInv")).getSpecialEnderChest(player, false);
how can i get an offline player inventory with this? becouse it only accepts Player and i cant cast OfflinePlayer to player

**Reply by Jikoo** [reply]
You use `IOpenInv#loadPlayer(OfflinePlayer)` to obtain a Player while they're offline. Note that all operations may not work (i.e. on older versions potion effect manipulation does not work), but inventory manipulation will. You don't need to use the `ISpecialInventory` constructs to clear inventories, those are specifically for user-friendly display of content.

**Comment by atesz19** [comment]
Thanks, but no luck :/
Can ypu help me?

```
for(OfflinePlayer p : Bukkit.getOfflinePlayers()){
            Player player = ((IOpenInv) plugin.getServer().getPluginManager().getPlugin("OpenInv")).loadPlayer(p);
            Inventory ec = player.getEnderChest();
            ec.remove(Material.SPAWNER);
            ec.close();
}
```
        
i want to remove all spawner items from players inventory and enderchest but this not working, can you able to help me please?

**Reply by Jikoo** [reply]
Well, for starters you're only accessing the ender chest, not the inventory. Note that Bukkit's regular inventory contents methods do not access armor/off hand slots for the `PlayerInventory`, so you'll also need to check those.

You probably will also want to call `IOpenInv#unload(OfflinePlayer)` after you're done so that your changes are saved to disk immediately.

**Comment by atesz19** [comment]
```
        Plugin openinv = plugin.getServer().getPluginManager().getPlugin("OpenInv");

        List<OfflinePlayer> offlist = new ArrayList<>(Arrays.asList(Bukkit.getOfflinePlayers()));

        for(OfflinePlayer p : offlist){

            Inventory inv = ((IOpenInv) openinv).loadPlayer(p).getInventory();
            inv.remove(Material.SPAWNER);
            inv.close();
            ((IOpenInv) openinv).unload(p);

        }
```

still not modify anything, whats the problem?
Also thank you very much for the help :)

**Reply by Jikoo** [reply]
Have you tried basic debugging? Are you checking the removed content to see if anything is actually changed? Try printing out things to make sure the code is working as expected.

You're now checking the player inventory and not the ender chest, I thought you wanted both? You're also not checking their off hand nor armor slots.
You can also manually call `Player#saveData` to ensure that data is saved immediately, forgot about that one.
